src/github.com/matrix-org/dendrite/clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,197 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/common"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// newMTag creates and returns a new MTag type
+func newMTag() common.MTag {
+	return common.MTag{
+		Tags: make(map[string]common.TagProperties),
+	}
+}
+
+// GetTag implements GET /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags
+func GetTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	userID string,
+	roomID string,
+) util.JSONResponse {
+	mtag := newMTag()
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+	err = json.Unmarshal(dataByte, &mtag)
+
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: mtag,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func PutTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	userID string,
+	roomID string,
+	tag string,
+) util.JSONResponse {
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+	var properties common.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	if len(data) > 0 {
+		dataByte, err := json.Marshal(data)
+		if err != nil {
+			httputil.LogThenError(req, err)
+		}
+		if err = json.Unmarshal(dataByte, &mtag); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+	mtag.Tags[tag] = properties
+	addDataToDB(req, localpart, roomID, accountDB, mtag)
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	userID string,
+	roomID string,
+	tag string,
+) util.JSONResponse {
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+
+	if len(data) > 0 {
+		dataByte, err := json.Marshal(data)
```

Comment:
Worth noting that whatever conclusion is reached on https://github.com/matrix-org/dendrite/pull/694/files#r267727079 should also be made here.